articles/modules/ROOT/pages/protecting-against-cypher-injection.adoc

```diff
@@ -38,119 +38,389 @@ Result result = session.run(queryString);
+For example: `node label with spaces`.
+This means that when dynamically building a query using string concatenation, sanitization needs to be done on the escaping of backticks.
+In Cypher, a backtick is escaped using another backtick `&#96;&#96;`.
+
+Here is an example of a simple dynamic label injection attack:
+
+`query = "MATCH (s:School)-[:IN]->(c:&#96;" + cityName + "&#96;) RETURN s;`
+
+With this query we want to search for all schools that are in a certain city, unfortunately our city names are Node Labels,
+so it isn't possible to parameterize the input.
+
+A possible attack input would be:
+
+`Input = &#96;) RETURN 1 as a UNION MATCH (n) RETURN 1 WITH DISTINCT true AS haxxored MATCH (n) DETACH DELETE n; \\`
+
+The backtick escapes the label name context and the parentheses closes the node.
+The `UNION` here then ensures that a match is made, as if the first `MATCH` statement doesn't return anything, the next part of the query won't be run.
+The `WITH` reduces the result set down to one row and then the final part will delete everything in the database.
+
+This attack was not possible to avoid using parameterization.
+To avoid this attack sanitization must be used.
+It is possible to add validation to the query as well, in this case, validating the city name is a real city name before passing it into the database, and rejecting it otherwise.
+
+The sanitization needed for this query is escaping the additional backtick character.
+
+`SanitizedInput = &#96;&#96;) RETURN 1 as a UNION MATCH (n) RETURN 1 WITH DISTINCT true AS haxxored MATCH (n) DETACH DELETE n; \\`
+
+The additional backtick added now ensures that the entire string is used as the node label, and not able to break out of that context.
+
+The unicode character for the bactick; `\u0060` will also resolve to a backtick and needs to be sanitized.
+When handling user input, it is important that the programming language the client is written in is taken into account.
+For example, the input: `\u005C\u00750060` may be resolved before being passed to the database as `\u0060`
+(`\u005C` is backslash, `\`, and `\u0075` is u), which will then be resolved by the database as a backtick!
+
+Writing your own sanitization function can be tricky.
+That is why it is highly recommended to avoid string concatenation and design your database in such a way that user input
+is not needed to dynamically query based on node labels, relationship types and parameters.
+
+It is also important to remember that not all injection attacks rely on quotation escaping of string literals,
+for example the following query, for a number literal, is still vulnerable when not using parameterization.
+
+`query = "MATCH (n) WHERE n.id =" + userid + ";"`
+
+Where the malicious userid input is: `"1 OR 1 = 1 WITH DISTINCT true AS haxxored MATCH (all) DETACH DELETE all; //"`
```

Comment:
This should probably be higher up, when discussing at a high level what type of input needs to be sanitised. Currently it is hidden away within the naming section.

Reply:
Fair enough!